presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -205,6 +205,70 @@ public ThriftMetastoreStats getStats()
         }
     }
 
+    @Override
+    public List<String> getTablesMatchingParameter(String databaseName, String parameterKey, String parameterValue)
+    {
+        try {
+            return retry()
+                    .stopOn(UnknownDBException.class)
+                    .stopOn(NoSuchObjectException.class)
+                    .stopOnIllegalExceptions()
+                    .run("getTablesMatchingParameter", stats.getGetTablesMatchingParameter().wrap(
+                            () -> doGetTablesMatchingParameter(databaseName, parameterKey, parameterValue)));
+        }
+        catch (UnknownDBException | NoSuchObjectException e) {
+            return ImmutableList.of();
+        }
+        catch (TException e) {
+            throw new PrestoException(HIVE_METASTORE_ERROR, e);
+        }
+        catch (Exception e) {
+            throw propagate(e);
+        }
+    }
+
+    private List<String> doGetTablesMatchingParameter(String databaseName, String parameterKey, String parameterValue)
+            throws TException
+    {
+        /*
+         * Thrift call `get_table_names_by_filter` may be translated by Metastore to a SQL query against Metastore database.
+         * Hive 2.3 on some databases uses CLOB for table parameter value column and some databases disallow `=` predicate over
+         * CLOB values. At the same time, they allow `LIKE` predicates over them.
+         */
+        String filterWithEquals = HIVE_FILTER_FIELD_PARAMS + parameterKey + " = \"" + parameterValue + "\"";
+        String filterWithLike = HIVE_FILTER_FIELD_PARAMS + parameterKey + " LIKE \"" + parameterValue + "\"";
```

Comment:
`_` and `%` need to be escaped in the value

if this metastore call doesn't support escaping, you can just fail on `parameterValue` containing these.

Reply:
Good catch.

I looked into the Hive code and found that the `parameterValue` here should be a Java regex in case of `LIKE` operator.
(https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java
http://www.datanucleus.org:15080/products/accessplatform_5_1/jdo/query.html#jdoql_methods)

I'd use `Pattern.quote()` here to escape metacharacters in Java.

Reply:
Uh... See https://github.com/apache/hive/blob/133d3c4733aca4fad9fd327bb14195aa5a78c664/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java#L61 (let me use `branch-3.1` as a reference for a moment)
`org.apache.hadoop.hive.metastore.parser.ExpressionTree.Operator#LIKE` gets converted to
- JDOQL `matches()` which is documented as "follows the rules of java.lang.String.matches method." (i.e. a `java.util.regex.Pattern`)
- SQL `like` which is a totally different thing.

From what I understand, which one is effective depends on Metastore configuration (whether it uses JDO or not).
Thus, if I'm not mistaken, the HMS behavior is hard to predict from outside.

Proposed solution:
- allow only parameter values that contain no regex special characters and no SQL LIKE special characters
  - for simplicity, let's limit this to parameter values matching `^[a-zA-Z0-9]*$`, we don't need anything else currently
- this requires some explanation in the code of course.

Reply:
Thank you for the explanation. Let's go with the proposed solution. I think it's sufficient for now.